### DISCUSSION TITLE
IKEv2 DPD timeout

### DISCUSSION THREAD

**Discussion by alexvelkov1** [strongswan/strongswan#840] [General]
Hi,

I have a IKEv2 IPsec tunnel between two boxes running the latest version of StrongSwan 5.9.4. I would like to configure DPD to check for peer connectivity and set a limit to clear the tunnel if not found active after a certain time (like the **dpdtimeout** parameter for IKEv1). 

I tried using the global **retransmit_limit** option in strongswan.conf, however charon behaves as if the option is not configured at all. Am I missing something, any ideas? 

Logs:

<pre>
BOX[~]# ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.4, Linux 5.10.27, armv7l):
  uptime: 51 minutes, since Jan 10 15:02:39 2022
  malloc: sbrk 184320, mmap 0, used 173376, free 10944
  worker threads: 27 of 32 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 6
  loaded plugins: charon gmp aes des sha2 sha3 sha1 md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey pem openssl fips-prf curve25519 xcbc hmac ccm gcm attr kernel-netlink resolve socket-default farp strok
Listening IP addresses:
  172.16.0.3
  10.0.3.1
Connections:
 ipsectest$0:  172.16.0.3...172.16.0.34  IKEv2, dpddelay=10s
 ipsectest$0:   local:  [client@company.com] uses pre-shared key authentication
 ipsectest$0:   remote: [server@company.com] uses pre-shared key authentication
 ipsectest$0:   child:  10.0.3.1/32 === 10.4.4.0/24 TUNNEL, dpdaction=clear
Routed Connections:
 ipsectest$0{1}:  ROUTED, TUNNEL, reqid 1
 ipsectest$0{1}:   10.0.3.1/32 === 10.4.4.0/24
Security Associations (1 up, 0 connecting):
 ipsectest$0[14]: ESTABLISHED 39 minutes ago, 172.16.0.3[client@company.com]...172.16.0.34[server@company.com]
 ipsectest$0[14]: IKEv2 SPIs: b320992d7dd13db8_i* ea70bc295ce2ce3d_r, rekeying in 21 minutes
 ipsectest$0[14]: IKE proposal: 3DES_CBC/HMAC_MD5_96/PRF_HMAC_MD5/MODP_1024
 ipsectest$0{3}:  INSTALLED, TUNNEL, reqid 1, ESP SPIs: cacab099_i ca0bbb3d_o
 ipsectest$0{3}:  3DES_CBC/HMAC_MD5_96, 0 bytes_i, 0 bytes_o, rekeying in 11 minutes
 ipsectest$0{3}:   10.0.3.1/32 === 10.4.4.0/24

BOX[~]# cat /var/etc/strongswan/strongswan.conf 
charon {
  ...
        threads=32
        retransmit_limit=60
   
        syslog {
                identifier=charon
                daemon {
                        ike=2
                        knl=2
                }
        }
}

libstrongswan {
        processor {
                priority_threads {
                        high = 1
                        medium = 4
                }
        }
}

BOX[~]#
Jan 10 15:54:07 BOX info  charon  [ 4271]: KNL 19     : querying policy 10.4.4.0/24 === 10.0.3.1/32 in
Jan 10 15:54:07 BOX info  charon  [ 4271]: KNL 19     : querying policy 10.4.4.0/24 === 10.0.3.1/32 fwd
Jan 10 15:54:07 BOX info  charon  [ 4271]: KNL 19     : querying SAD entry with SPI cacab099
Jan 10 15:54:07 BOX info  charon  [ 4271]: IKE 19     : sending DPD request
Jan 10 15:54:07 BOX info  charon  [ 4271]: IKE 19     : queueing IKE_MOBIKE task
Jan 10 15:54:07 BOX info  charon  [ 4271]: IKE 19     : activating new tasks
Jan 10 15:54:07 BOX info  charon  [ 4271]: IKE 19     :   activating IKE_MOBIKE task
Jan 10 15:54:07 BOX info  charon  [ 4271]: KNL 19     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:07 BOX info  charon  [ 4271]: ENC 19     : generating INFORMATIONAL request 240 [ ]
Jan 10 15:54:07 BOX info  charon  [ 4271]: KNL 19     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:07 BOX info  charon  [ 4271]: NET 19     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)
Jan 10 15:54:07 BOX info  charon  [ 4271]: NET 15     : received packet: from 172.16.0.34[4500] to 172.16.0.3[4500] (60 bytes)
Jan 10 15:54:07 BOX info  charon  [ 4271]: ENC 15     : parsed INFORMATIONAL response 240 [ ]
Jan 10 15:54:07 BOX info  charon  [ 4271]: IKE 15     : activating new tasks
Jan 10 15:54:07 BOX info  charon  [ 4271]: IKE 15     : nothing to initiate
Jan 10 15:54:08 BOX info  charon  [ 4271]: KNL 12     : querying SAD entry with SPI cacab099
Jan 10 15:54:08 BOX info  charon  [ 4271]: KNL 12     : querying SAD entry with SPI ca0bbb3d
Jan 10 15:54:17 BOX info  charon  [ 4271]: KNL 16     : querying policy 10.4.4.0/24 === 10.0.3.1/32 in
Jan 10 15:54:17 BOX info  charon  [ 4271]: KNL 16     : querying policy 10.4.4.0/24 === 10.0.3.1/32 fwd
Jan 10 15:54:17 BOX info  charon  [ 4271]: KNL 16     : querying SAD entry with SPI cacab099
Jan 10 15:54:17 BOX info  charon  [ 4271]: IKE 16     : sending DPD request
Jan 10 15:54:17 BOX info  charon  [ 4271]: IKE 16     : queueing IKE_MOBIKE task
Jan 10 15:54:17 BOX info  charon  [ 4271]: IKE 16     : activating new tasks
Jan 10 15:54:17 BOX info  charon  [ 4271]: IKE 16     :   activating IKE_MOBIKE task
Jan 10 15:54:17 BOX info  charon  [ 4271]: KNL 16     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:17 BOX info  charon  [ 4271]: ENC 16     : generating INFORMATIONAL request 241 [ ]
Jan 10 15:54:17 BOX info  charon  [ 4271]: KNL 16     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:17 BOX info  charon  [ 4271]: NET 16     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)
Jan 10 15:54:17 BOX info  charon  [ 4271]: NET 21     : received packet: from 172.16.0.34[4500] to 172.16.0.3[4500] (60 bytes)
Jan 10 15:54:17 BOX info  charon  [ 4271]: ENC 21     : parsed INFORMATIONAL response 241 [ ]
Jan 10 15:54:17 BOX info  charon  [ 4271]: IKE 21     : activating new tasks
Jan 10 15:54:17 BOX info  charon  [ 4271]: IKE 21     : nothing to initiate

### block all traffic on 172.16.0.34 (the peer) coming from 172.16.0.3 (this machine) ###

Jan 10 15:54:27 BOX info  charon  [ 4271]: KNL 22     : querying policy 10.4.4.0/24 === 10.0.3.1/32 in
Jan 10 15:54:27 BOX info  charon  [ 4271]: KNL 22     : querying policy 10.4.4.0/24 === 10.0.3.1/32 fwd
Jan 10 15:54:27 BOX info  charon  [ 4271]: KNL 22     : querying SAD entry with SPI cacab099
Jan 10 15:54:27 BOX info  charon  [ 4271]: IKE 22     : sending DPD request
Jan 10 15:54:27 BOX info  charon  [ 4271]: IKE 22     : queueing IKE_MOBIKE task
Jan 10 15:54:27 BOX info  charon  [ 4271]: IKE 22     : activating new tasks
Jan 10 15:54:27 BOX info  charon  [ 4271]: IKE 22     :   activating IKE_MOBIKE task
Jan 10 15:54:27 BOX info  charon  [ 4271]: KNL 22     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:27 BOX info  charon  [ 4271]: ENC 22     : generating INFORMATIONAL request 242 [ ]
Jan 10 15:54:27 BOX info  charon  [ 4271]: KNL 22     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:27 BOX info  charon  [ 4271]: NET 22     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)
Jan 10 15:54:31 BOX info  charon  [ 4271]: IKE 06     : retransmit 1 of request with message ID 242
Jan 10 15:54:31 BOX info  charon  [ 4271]: KNL 06     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:31 BOX info  charon  [ 4271]: NET 06     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)
Jan 10 15:54:39 BOX info  charon  [ 4271]: IKE 07     : retransmit 2 of request with message ID 242
Jan 10 15:54:39 BOX info  charon  [ 4271]: KNL 07     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:39 BOX info  charon  [ 4271]: NET 07     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)
Jan 10 15:54:52 BOX info  charon  [ 4271]: IKE 08     : retransmit 3 of request with message ID 242
Jan 10 15:54:52 BOX info  charon  [ 4271]: KNL 08     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:54:52 BOX info  charon  [ 4271]: NET 08     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)
Jan 10 15:55:15 BOX info  charon  [ 4271]: IKE 18     : retransmit 4 of request with message ID 242
Jan 10 15:55:15 BOX info  charon  [ 4271]: KNL 18     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:55:15 BOX info  charon  [ 4271]: NET 18     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)

### at around 15:55:27 the transmission limit was reached, but charon continues the transmission before finally giving up at 15:56:57

Jan 10 15:55:57 BOX info  charon  [ 4271]: IKE 27     : retransmit 5 of request with message ID 242
Jan 10 15:55:57 BOX info  charon  [ 4271]: KNL 27     : using 172.16.0.3 as address to reach 172.16.0.34/32
Jan 10 15:55:57 BOX info  charon  [ 4271]: NET 27     : sending packet: from 172.16.0.3[4500] to 172.16.0.34[4500] (60 bytes)
Jan 10 15:56:57 BOX info  charon  [ 4271]: IKE 18     : giving up after 5 retransmits
Jan 10 15:56:57 BOX info  charon  [ 4271]: IKE 18     : IKE_SA ipsectest$0[14] state change: ESTABLISHED => DESTROYING
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : deleting policy 10.0.3.1/32 === 10.4.4.0/24 out
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : policy still used by another CHILD_SA, not removed
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : updating policy 10.0.3.1/32 === 10.4.4.0/24 out [priority 371328, refcount 1]
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : deleting policy 10.4.4.0/24 === 10.0.3.1/32 in
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : policy still used by another CHILD_SA, not removed
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : updating policy 10.4.4.0/24 === 10.0.3.1/32 in [priority 371328, refcount 1]
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : deleting policy 10.4.4.0/24 === 10.0.3.1/32 fwd
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : policy still used by another CHILD_SA, not removed
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : updating policy 10.4.4.0/24 === 10.0.3.1/32 fwd [priority 371328, refcount 1]
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : deleting SAD entry with SPI cacab099
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : deleted SAD entry with SPI cacab099
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : deleting SAD entry with SPI ca0bbb3d
Jan 10 15:56:57 BOX info  charon  [ 4271]: KNL 18     : deleted SAD entry with SPI ca0bbb3d                                                                                                                                                                                          
</pre>

**Comment by tobiasbrunner** [comment]
I think you misunderstood what that option does, please read [Retransmission](https://wiki.strongswan.org/projects/strongswan/wiki/Retransmission).

**Comment by alexvelkov1** [comment]
Hi Tobias,

thanks for your answer!

You are right. I had already read the Retransmission page before and I thought **retransmit_limit** is the upper-limit for the resulting overall timeout, but it is only for the time between the retransmits.
So, I had a look in the logs again and I see that it took exactly 60s between "retransmit 5.." and "giving up..", so I guess, this is where the **retransmit_limit** operates.
